### ISSUE TITLE
maximum call stack size exceeded

### ISSUE THREAD

**Issue by codeback** [leviwheatcroft/superagent-throttle#14]
When we do high volume of requests (more than 1000)  we are receiving the following error :
```
timers.js:453
function Timeout(after, callback, args) {
                ^

RangeError: Maximum call stack size exceeded
    at new Timeout (timers.js:453:17)
    at createSingleTimeout (timers.js:351:15)
    at exports.setTimeout (timers.js:343:10)
    at Throttle.cycle (/home/workspace/conector/node_modules/superagent-throttle/dist/index.js:195:27)
    at Request.cleanup [as _maskedCallback] (/home/workspace/conector/node_modules/superagent-throttle/dist/index.js:225:16)
    at Request.cleanup [as _maskedCallback] (/home/workspace/conector/node_modules/superagent-throttle/dist/index.js:227:15)
    at Request.cleanup [as _maskedCallback] (/home/workspace/conector/node_modules/superagent-throttle/dist/index.js:227:15)
    at Request.cleanup [as _maskedCallback] (/home/workspace/conector/node_modules/superagent-throttle/dist/index.js:227:15)
    at Request.cleanup [as _maskedCallback] (/home/workspace/conector/node_modules/superagent-throttle/dist/index.js:227:15)
    at Request.cleanup [as _maskedCallback] (/home/workspace/conector/node_modules/superagent-throttle/dist/index.js:227:15)
    at Request.cleanup [as _maskedCallback] (/home/workspace/conector/node_modules/superagent-throttle/dist/index.js:227:15)
    at Request.cleanup [as _maskedCallback] (/home/workspace/conector/node_modules/superagent-throttle/dist/index.js:227:15)
    at Request.cleanup [as _maskedCallback] (/home/workspace/conector/node_modules/superagent-throttle/dist/index.js:227:15)
    at Request.cleanup [as _maskedCallback] (/home/workspace/conector/node_modules/superagent-throttle/dist/index.js:227:15)
    at Request.cleanup [as _maskedCallback] (/home/workspace/conector/node_modules/superagent-throttle/dist/index.js:227:15)
    at Request.cleanup [as _maskedCallback] (/home/workspace/conector/node_modules/superagent-throttle/dist/index.js:227:15)
```
This error only happens if we use the plugin. 

Any idea of the problem? Thanks.

**Comment by leviwheatcroft** [comment]
Seems odd to have multiple `_maskedCallback`'s in your stack.

Can you post your instantiation and usage? I don't need working code, just want to take a look at your use case.

**Comment by codeback** [comment]
Hi @leviwheatcroft, thanks for your quick response. 

Here I make the request:

```
...
function createProducts(searchUrl) {
  Request.get(searchUrl)
  .then((parsedHTML) => {
    findProductUrls(parsedHTML).map((url) => newPage.emit(Request.get(searchUrl)));
    return parsedHTML;
  })
  .then((parsedHTML) => checkNext(parsedHTML))
  .catch((err) => {
    throw err;
  });
}

function checkNext(parsedHTML) {
  let hasNext = findNextPage(parsedHTML);
  if (lib.isThere(hasNext)) {
    createProducts(hasNext);
  } else {
    newPage.end();
  }
}

createProducts(searchPageUrl);
```

And here is the Request:
```
class Request {

  constructor(params) {

    this.throttle = params ?
      new Throttle(params) :
      null;

  }

  get(urlStr) {

    return new Promise((resolve,reject) => {
      let req =  this.throttle ?
        request.get(encodeURI(urlStr)).use(this.throttle.plugin()) :
        request.get(encodeURI(urlStr));

      return req.end(function(err, res) {

          // Do stuff

        });
    });

  }

}

module.exports = Request;

```

I hope this help.

**Comment by leviwheatcroft** [comment]
Is it possible you're creating a new instance of throttle for every request ? I don't see where you've instantiated Request as in `let request = new Request({...})`. I'm a bit confused by `Request.get()` instead of `request.get()` in `createProducts`.

Anyhow, if you are instantiating Request 1000 times then you'll have 1000 throttles which would explain the excessive stack.

This may well be a bug, but the "should work when resource bound" test can be wound up to queue 5000 requests and issue them 1000 at a time without issue.

**Comment by codeback** [comment]
No, I am creating only one instance. I do in this way:  
```
let Request = new RequestClass(throttleParams);
```

> This may well be a bug, but the "should work when resource bound" test can be wound up to queue 5000 requests and issue them 1000 at a time without issue.

Yes it is strange, but if I remove the plugin, it works. :(

**Comment by leviwheatcroft** [comment]
Ok, well none of the tests use promises i guess.. so i could include something like that and see if it reveals anything. 

Probably won't be within 48 hours.

**Comment by codeback** [comment]
Ok, thank you very much!!

**Comment by wasadigi** [comment]
I am hitting the same problem. I did not fully grasp what is happening, but I suspect that when the server sends a REDIRECT (301), the issue arises. In my test, I am accessing a REST API via HTTP and get re-directed to HTTPS.

**Comment by codeback** [comment]
Hi @wasadigi, I can confirm that in my case, it happens when the server sends a the 301 code too. 

@leviwheatcroft can it be the problem?

**Comment by leviwheatcroft** [comment]
Thanks for narrowing this down for me, I trully appreciate it.

I've had a quick look but the problem isn't obviou8s.

Sorry but I just don't have time to resolving this at the moment, maybe late next week but no promises.

**Comment by Finesse** [comment]
I also have a redirect and I get this error too:

```js
const throttle = new SuperagentThrottle({
	rate: 3,
	ratePer: 1000,
	concurrent: 3
});

superagent
	.get('http://myserver.local/redirect')
	.use(throttle.plugin())
	.then(response => console.log(response));
```

But when I use SuperAgent this way, nothing happens (neither a success, nor an error):

```js
const throttle = new SuperagentThrottle({
	rate: 1,
	ratePer: 1000,
	concurrent: 1
});

superagent
	.get('http://myserver.local/redirect')
	.use(throttle.plugin())
	.then(response => console.log(response));
```

**Comment by leviwheatcroft** [comment]
So, I get that this is 18 months late, and that posters in this thread have likely long since moved on, however I've finally tracked this down and written a test that throws the error described.

More for my own reference than anything, superagent handles a redirect by [calling the `end` method again](https://github.com/visionmedia/superagent/blob/3b25a431e23d63cec9abd9957a280a51c3691a7a/lib/node/index.js#L519). Due to `superagent-throttles` pattern of masking the `end` method, you end up in a loop.

Sadly, at this point an appropriate fix is unclear. I guess this is a pretty good example of why wrapping methods on objects from other libraries is a bad idea, because it's difficult to be sure you cover all the uses of the method you're wrapping.

**Comment by leviwheatcroft** [comment]
Just published v1.0.1 which fixes this.
